scapy/contrib/automotive/xcp/xcp.py

```diff
@@ -0,0 +1,462 @@
+
+conf.contribs['XCP']['timestamp_size'] = 0
+
+
+# Specifications from:
+# http://read.pudn.com/downloads293/doc/comm/1316424/ASAM_XCP_Part1-Overview_V1.0.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%202-%20Protocol%20Layer%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%203-%20Transport_layer_specification_xcp_on_can_1-0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%204-%20Interface%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%205-%20Example%20Communication%20Sequences%20-1.0.pdf # noqa: E501
+
+# XCP on USB is left out because it has "no practical meaning"
+# XCP on Lin is left out because it has no official specification
+class XCPOnCAN(CAN):
+    name = "Universal calibration and measurement protocol on CAN"
+    fields_desc = [
+        FlagsField("flags", 0, 3, ["error",
+                                   "remote_transmission_request",
+                                   "extended"]),
+        XBitField("identifier", 0, 29),
+        ByteField("length", None),
+        ThreeBytesField("reserved", 0),
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.length is None:
+            tmp_len = len(pay)
+            pkt = pkt[:4] + struct.pack("B", tmp_len) + pkt[5:]
+        return super(XCPOnCAN, self).post_build(pkt, pay)
+
+    def extract_padding(self, p):
+        return p, None
+
+
+class XCPOnUDP(UDP):
+    name = "Universal calibration and measurement protocol on Ethernet"
+    fields_desc = UDP.fields_desc + [
+        ShortField("length", None),
+        ShortField("ctr", 0),  # counter
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.length is None:
+            tmp_len = len(pay)
+            pkt = pkt[:8] + struct.pack("!H", tmp_len) + pkt[10:]
+        return super(XCPOnUDP, self).post_build(pkt, pay)
+
+
+class XCPOnTCP(TCP):
+    name = "Universal calibration and measurement protocol on Ethernet"
+
+    fields_desc = TCP.fields_desc + [
+        ShortField("length", None),
+        ShortField("ctr", 0),  # counter
+    ]
+
+    def answers(self, other):
+        if not isinstance(other, XCPOnTCP):
+            return 0
+        if isinstance(other.payload, CTORequest) and isinstance(self.payload,
+                                                                CTOResponse):
+            return self.payload.answers(other.payload)
+
+    def post_build(self, pkt, pay):
+        if self.length is None:
+            len_offset = 20 + len(self.options)
+            tmp_len = len(pay)
+            tmp_len = struct.pack("!H", tmp_len)
+            pkt = pkt[:len_offset] + tmp_len + pkt[len_offset + 2:]
+        return super(XCPOnTCP, self).post_build(pkt, pay)
+
+
+class XCPOnCANTail(Packet):
+    name = "XCP Tail on CAN"
+
+    fields_desc = [
+        StrField("control_field", "")
+    ]
+
+
+class CTORequest(Packet):
```

Comment:
This class may need a answers function

Reply:
Why? I think this will never be a response.